### PR TITLE
Add extra DICOM and ITK metadata as PanImg fields

### PR DESCRIPTION
**Pull request by HarmvZ**
**Note: Merges into feature branch**

Adds optional metadata from dicom and itk headers as fields of `PanImg` objects for the following keys:
https://github.com/DIAGNijmegen/rse-panimg/blob/b46606818d345fd9783a70b9268258bea2f56f9f/panimg/models.py#L98-L106

It performs some loose validation on the values and casts them to the correct type specified here:
https://github.com/DIAGNijmegen/rse-panimg/blob/b46606818d345fd9783a70b9268258bea2f56f9f/panimg/models.py#L126-L134

Validation is (only) done in the builders before the images are converted to the intermediate `SimpleITKImage` object. It will raise a `ValidationError` if any of the optional headers have an incorrect format. This means the image is skipped.

### CODE REVIEW

File: panimg/image_builders/dicom.py [hunk 1]
```diff
@@ -259,7 +251,11 @@ def _process_dicom_file(*, dicom_ds):  # noqa: C901
 
     for f in OPTIONAL_METADATA_FIELDS:
         if hasattr(ref_file, f):
-            img.SetMetaData(f, str(getattr(ref_file, f)))
+            value = getattr(ref_file, f)
+            key_to_md = {md.keyword: md for md in EXTRA_METADATA}
+            if f in key_to_md:
+                key_to_md[f].validate_value(value)
```
Comment (jmsmkn) [inline panimg/image_builders/dicom.py:257]: This construction is a little confusing, re-constructed for each item in the header and duplicated elsewhere in the code. Could it instead be moved to a method?

File: panimg/models.py [hunk 2]
```diff
@@ -39,6 +43,70 @@ class EyeChoice(str, Enum):
...
+
+DICOM_VR_TO_VALIDATION_REGEXP = {
+    "AS": re.compile(r"^\d{3}[DWMY]$"),
+    "CS(PatientSex)": re.compile(
+        f"^[{''.join([s.value for s in PatientSex])}]$"
+    ),
+    "DA": re.compile(r"^\d{8}$"),
+    "LO": re.compile(r"^[^\\]{0,64}$"),
+    "PN": re.compile(r"^[^\\]{0,324}$"),
+    "UI": re.compile(r"^[\d.]{0,64}$"),
```
Comment (HarmvZ) [inline panimg/models.py:60]: I'm not completely sure about the strictness of this validation. It's for value representation `UI` (=StudyInstanceUID, etc). Right now it only allows digits and "." as specified in the DICOM standard. For the ophthalmology usecase a study descriptor like `Visit 6` is required. So the `StudyInstanceUID` field can not be used for that currently. The `StudyDescription` field can be used for it but I'm not sure if that is the right one.

File: panimg/models.py [hunk 3]
```diff
@@ -39,6 +43,70 @@ class EyeChoice(str, Enum):
...
+class PatientSex(str, Enum):
+    MALE = "M"
+    FEMALE = "F"
+    OTHER = "O"
+
+
+DICOM_VR_TO_VALIDATION_REGEXP = {
+    "AS": re.compile(r"^\d{3}[DWMY]$"),
+    "CS(PatientSex)": re.compile(
+        f"^[{''.join([s.value for s in PatientSex])}]$"
```
Comment (jmsmkn) [inline panimg/models.py:55]: Isn't this being done in the enum validation?

Comment (HarmvZ) [reply panimg/models.py]: Just checked and indeed pydantic takes care of it. Nice!

File: panimg/models.py (hunk 2 above)
Comment (jmsmkn) [reply panimg/models.py:58]: We discussed this in the RSE meeting and decided that the StudyDescription is the correct place for this. Any validation that panimg does for these fields should only be according to the fields value representation http://dicom.nema.org/dicom/2013/output/chtml/part05/sect_6.2.html

**Comment by HarmvZ** [comment]
As mentioned in https://github.com/comic/grand-challenge.org/pull/2051#discussion_r717627141, an extra validation step is added in between `SimpleITKImage` -> `PanImg` conversion to check if the values for the metadata adheres to the DICOM standard value representations.

Failing the validation there will strip (=not add) the specific metadata field from the object (not from the actual file) and it throws a warning. The actual validation of the values should and currently is already be done in the builder step. There it is possible to throw a ValidationError that results in the builder skipping only that file and continuing the import session. This is not possible in this extra validation step. Throwing an error there would result in the whole import session failing, so the choice is made to strip the value and log a warning. I also added a test for this.

**Review by jmsmkn — APPROVED** [review]

**Pull request merged**